In [2]:
import pandas as pd
from scipy.stats import chi2_contingency, ttest_ind

In [ ]:
df = pd.read_csv("../data/engineered/BankChurners_Engineered.csv")
df.head()

Attrition_Flag  Customer_Age Gender  Dependent_count Marital_Status  \
0  Existing Customer            45      M                3        Married   
1  Existing Customer            49      F                5         Single   
2  Existing Customer            51      M                3        Married   
3  Existing Customer            40      F                4        Unknown   
4  Existing Customer            40      M                3        Married   

  Card_Category  Months_on_book  Total_Relationship_Count  \
0          Blue              39                         5   
1          Blue              44                         6   
2          Blue              36                         4   
3          Blue              34                         3   
4          Blue              21                         5   

   Months_Inactive_12_mon  Contacts_Count_12_mon  ...  Total_Ct_Chng_Q4_Q1  \
0                       1                      3  ...                1.625   
1                       1                      2  ...                3.714   
2                       1                      0  ...                2.333   
3                       4                      1  ...                2.333   
4                       1                      0  ...                2.500   

   Avg_Utilization_Ratio  Income_Start  Income_End  Education_Level_Encoded  \
0                  0.061       60000.0     80000.0                        1   
1                  0.105           0.0     40000.0                        3   
2                  0.000       80000.0    120000.0                        3   
3                  0.760           0.0     40000.0                        1   
4                  0.000       60000.0     80000.0                        0   

   Activity_Ratio  Engagement_Score  High_Credit_Flag  Trans_Freq_Group  \
0       27.238095                84                 1               Low   
1       39.121212                79                 1               Low   
2       94.350000                56                 0               Low   
3       58.550000                55                 0               Low   
4       29.142857                49                 1               Low   

   Tenure_Group  
0        Medium  
1        Medium  
2        Medium  
3        Medium  
4           New  

[5 rows x 26 columns]

In [21]:
# If p < 0.05, Gender has a statistically significant relationship with churn.

table = pd.crosstab(df["Gender"], df["Attrition_Flag"])
chi2, p, dof, exp = chi2_contingency(table)

print("p_value:", p)
#-------------------------------------------------------------------------------


table = pd.crosstab(df["Education_Level_Encoded"], df["Attrition_Flag"])
chi2, p, dof, exp = chi2_contingency(table)

print("p-value:", p)
#-------------------------------------------------------------------------------

table = pd.crosstab(df["Marital_Status"], df["Attrition_Flag"])
chi2, p, dof, exp = chi2_contingency(table)

print("p-value:", p)
#-------------------------------------------------------------------------------

table = pd.crosstab(df["Card_Category"], df["Attrition_Flag"])
chi2, p, dof, exp = chi2_contingency(table)

print("p-value:", p)
#-------------------------------------------------------------------------------

table = pd.crosstab(df["Tenure_Group"], df["Attrition_Flag"])
chi2, p, dof, exp = chi2_contingency(table)

print("p-value:", p)

p_value: 0.00019635846717310307
p-value: 0.05148913147336627
p-value: 0.10891263394840227
p-value: 0.5252382797994759
p-value: 0.526762810172551


🔍 **Statistical Finding (Chi-Square Test)**  
There is a statistically significant relationship between **Gender** and customer churn  
(p = 0.0002 < 0.05).  

This indicates that churn behavior differs between male and female customers.

🔍 **Statistical Finding (Chi-Square Test)**  
Education level shows a **borderline relationship** with customer churn  
(p ≈ 0.051).  

Since the p-value is slightly above 0.05, education level is **not strongly associated**
with churn at the 95% confidence level.


🔍 **Statistical Finding (Chi-Square Test)**  
Marital status does not have a statistically significant relationship with churn  
(p = 0.109 > 0.05).  

This suggests that marital status alone does not meaningfully influence customer churn.


🔍 **Statistical Finding (Chi-Square Test)**  
Card category is not significantly associated with customer churn  
(p = 0.525 > 0.05).  

Customers with different card categories exhibit similar churn behavior.


🔍 **Statistical Finding (Chi-Square Test)**  
Customer tenure group does not show a statistically significant relationship with churn  
(p = 0.527 > 0.05).  

This implies that churn is not solely determined by how long a customer has been with the bank.


In [20]:
churned = df[df["Attrition_Flag"] == "Attrited Customer"]["Credit_Limit"]
existing = df[df["Attrition_Flag"] == "Existing Customer"]["Credit_Limit"]

t_stat, p_val = ttest_ind(churned, existing)
print("p-value:", p_val)

#------------------------------------------------------------------------------------------


churned = df[df["Attrition_Flag"] == "Attrited Customer"]["Total_Trans_Amt"]
existing = df[df["Attrition_Flag"] == "Existing Customer"]["Total_Trans_Amt"]

t_stat, p_val = ttest_ind(churned, existing)
print("p-value:", p_val)

#------------------------------------------------------------------------------------------

churned = df[df["Attrition_Flag"] == "Attrited Customer"]["Activity_Ratio"]
existing = df[df["Attrition_Flag"] == "Existing Customer"]["Activity_Ratio"]

t_stat, p_val = ttest_ind(churned, existing)
print("p-value:", p_val)

#------------------------------------------------------------------------------------------

churned = df[df["Attrition_Flag"] == "Attrited Customer"]["Engagement_Score"]
existing = df[df["Attrition_Flag"] == "Existing Customer"]["Engagement_Score"]

t_stat, p_val = ttest_ind(churned, existing)
print("p-value:", p_val)

p-value: 0.016285357205394337
p-value: 1.857438655661051e-65
p-value: 0.10209849132530774
p-value: 1.4934122829496365e-281


🔍 **Statistical Finding (T-Test)**  
There is a statistically significant difference in **credit limit** between churned and
existing customers (p = 0.016 < 0.05).  

This suggests that customers with lower credit limits are more likely to churn.


🔍 **Statistical Finding (T-Test)**  
Total transaction amount differs significantly between churned and existing customers  
(p < 0.001).  

Customers who churn tend to have **much lower transaction amounts**, indicating reduced
financial activity before churn.

🔍 **Statistical Finding (T-Test)**  
There is no statistically significant difference in **activity ratio** between churned
and non-churned customers (p = 0.102 > 0.05).  

This feature alone may not be a strong predictor of churn.

🔍 **Statistical Finding (T-Test)**  
Engagement score shows a highly significant difference between churned and existing customers  
(p < 0.001).  

Customers who churn have **substantially lower engagement scores**, confirming that reduced
interaction with the bank strongly predicts churn.

## Overall Statistical Insights

The statistical analysis reveals that **customer engagement and transaction behavior**
are the strongest indicators of churn.

While demographic factors such as marital status and card category show limited influence,
behavioral features like **transaction amount** and **engagement score** are highly
significant predictors.

These findings justify the use of engineered behavioral features in the predictive
modeling stage.
